kvbc/include/categorization/details.h

```diff
@@ -0,0 +1,74 @@
+// Concord
+//
+// Copyright (c) 2020 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the
+// "License").  You may not use this product except in compliance with the
+// Apache 2.0 License.
+//
+// This product may include a number of subcomponents with separate copyright
+// notices and license terms. Your use of these subcomponents is subject to the
+// terms and conditions of the subcomponent's license, as noted in the LICENSE
+// file.
+
+#pragma once
+
+#include "kv_types.hpp"
+#include "sparse_merkle/base_types.h"
+
+#include <optional>
+#include <set>
+#include <unordered_map>
+#include <variant>
+
+namespace concord::kvbc::categorization::detail {
+
+struct MerkleUpdateInfo {
+  struct KeyFlags {
+    bool deleted{false};
+  };
+
+  std::unordered_map<Key, KeyFlags> keys;
+  std::array<std::uint8_t, 32> hash;
+  sparse_merkle::Version version;
+  std::string category_id;
+
+  template <typename T>
+  bool operator<(const T &other) const {
+    return category_id < other.category_id;
+  }
+};
+
+struct KeyValueUpdateInfo {
+  struct KeyFlags {
+    bool deleted{false};
+
+    // Applicable to non-deleted keys only.
+    bool stale_on_update{false};
+  };
```

Comment:
Does this bool should be part of the Block as well i.e. the Node for this category will  contain a map<key, KeyFlags >

Reply:
I think yes, because of State Transfer. We should be able to do the exact same operations (marking stale) on the receiving replica. Please correct me if I miss something.

Reply:
I guess that if the stale flag is not part of the value struct that is stored in DB then it's essential.

Anyway, it got me thinking about ST in general, and currently, the raw block enables us to call the Merkle add block as any other client would.
We should ensure that the new Raw block will contain all the information needed to construct the per group update struct as defined in updates.h

Reply:
Yes, agreed.